server/src/main/java/org/elasticsearch/index/codec/vectors/DefaultIVFVectorsReader.java

```diff
@@ -0,0 +1,420 @@
+
+import java.io.IOException;
+import java.util.function.IntPredicate;
+
+import static org.apache.lucene.codecs.lucene102.Lucene102BinaryQuantizedVectorsFormat.QUERY_BITS;
+import static org.apache.lucene.index.VectorSimilarityFunction.COSINE;
+import static org.apache.lucene.index.VectorSimilarityFunction.EUCLIDEAN;
+import static org.apache.lucene.index.VectorSimilarityFunction.MAXIMUM_INNER_PRODUCT;
+import static org.apache.lucene.util.quantization.OptimizedScalarQuantizer.discretize;
+import static org.apache.lucene.util.quantization.OptimizedScalarQuantizer.transposeHalfByte;
+import static org.elasticsearch.simdvec.ES91OSQVectorsScorer.BULK_SIZE;
+
+/**
+ * Default implementation of {@link IVFVectorsReader}. It scores the posting lists centroids using
+ * brute force and then scores the top ones using the posting list.
+ */
+public class DefaultIVFVectorsReader extends IVFVectorsReader {
+    private static final float FOUR_BIT_SCALE = 1f / ((1 << 4) - 1);
+
+    public DefaultIVFVectorsReader(SegmentReadState state, FlatVectorsReader rawVectorsReader) throws IOException {
+        super(state, rawVectorsReader);
+    }
+
+    @Override
+    CentroidQueryScorer getCentroidScorer(
+        FieldInfo fieldInfo,
+        int numCentroids,
+        IndexInput centroids,
+        float[] targetQuery,
+        IndexInput clusters
+    ) throws IOException {
+        FieldEntry fieldEntry = fields.get(fieldInfo.number);
+        float[] globalCentroid = fieldEntry.globalCentroid();
+        float globalCentroidDp = fieldEntry.globalCentroidDp();
+        OptimizedScalarQuantizer scalarQuantizer = new OptimizedScalarQuantizer(fieldInfo.getVectorSimilarityFunction());
+        byte[] quantized = new byte[targetQuery.length];
+        float[] targetScratch = ArrayUtil.copyArray(targetQuery);
+        OptimizedScalarQuantizer.QuantizationResult queryParams = scalarQuantizer.scalarQuantize(
+            targetScratch,
+            quantized,
+            (byte) 4,
+            globalCentroid
+        );
+        return new CentroidQueryScorer() {
+            int currentCentroid = -1;
+            private final byte[] quantizedCentroid = new byte[fieldInfo.getVectorDimension()];
+            private final float[] centroid = new float[fieldInfo.getVectorDimension()];
+            private final float[] centroidCorrectiveValues = new float[3];
+            private int quantizedCentroidComponentSum;
+            private final long centroidByteSize = fieldInfo.getVectorDimension() + 3 * Float.BYTES + Short.BYTES;
+
+            @Override
+            public int size() {
+                return numCentroids;
+            }
+
+            @Override
+            public float[] centroid(int centroidOrdinal) throws IOException {
+                readQuantizedCentroid(centroidOrdinal);
+                return centroid;
+            }
+
+            private void readQuantizedCentroid(int centroidOrdinal) throws IOException {
```

Comment:
the function name is `readQuantizedCentroid` but it also reads full centroid. It looks like the second read of full centroid is not necessary in scoring? Should we break this function be only about reading quantized centroid?